### PR TITLE
Triplets

### PR DESCRIPTION
**Pull request by fenallen**
What

Added a v3 input file parser and it test to load files and hierarchies.
Some it tests are temporarily disabled whilst we regenerate input hierarchy files with latest naming changes.

How to review

Checkout the corresponding backend model code and run the it tests.

Who can review

Anyone by especially @YetAnotherMatt

### CODE REVIEW

File: app/services/InputCSVParserV3.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import exceptions.GLLoadException;
+import play.Logger;
+import uk.co.onsdigital.discovery.model.DataPoint;
+import uk.co.onsdigital.discovery.model.DimensionValue;
+import uk.co.onsdigital.discovery.model.DimensionalDataSet;
+import uk.co.onsdigital.discovery.model.HierarchyEntry;
+
+import javax.persistence.EntityManager;
+
+import java.io.*;
```
Comment (YetAnotherMatt) [inline app/services/InputCSVParserV3.java:12]: Star imports

File: app/services/InputCSVParserV3.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        ArrayList<DimensionValue> dimensions = new ArrayList<>();
+
+        for (int i = 3; i < rowData.length; i = i + 3) {
+            String hierarchyId = rowData[i];
+            String dimensionName = rowData[i + 1];
+            String dimensionValue = rowData[i + 2];
+
+            logger.debug("Creating dimension for hierarchyId: " + hierarchyId + " and dimensionName: " + dimensionName + " and dimensionValue: " + dimensionValue + " ....");
```
Comment (YetAnotherMatt) [inline app/services/InputCSVParserV3.java:41]: Should be:
logger.debug("Creating dimension for hierarchyId: {}  and dimensionName: {}and dimensionValue: {} ....",  hierarchyId, dimensionName, dimensionValue);

File: app/services/InputCSVParserV3.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            if(existing.isEmpty()) {
+                em.persist(dimension);
+            }
+
+            DimensionValue newDimension = em.createQuery("SELECT dim FROM DimensionValue dim WHERE dim.dimensionalDataSetId = :ddsId AND dim.name = :name AND dim.value = :value", DimensionValue.class)
+                    .setParameter("ddsId", dds.getId())
+                    .setParameter("name", dimensionName)
+                    .setParameter("value", dimensionValue)
+                    .getSingleResult();
+            dimensions.add(newDimension);
```
Comment (YetAnotherMatt) [inline app/services/InputCSVParserV3.java:70]: Do we need to run this query twice? Can't we just use either the result from existing, or the persisted dimension?

File: app/services/InputCSVParserV3.java (hunk 1 above)
Comment (fenallen) [reply app/services/InputCSVParserV3.java:12]: ...are completely fine in the 21st century.

File: app/services/InputCSVParserV3.java (hunk 2 above)
Comment (fenallen) [reply app/services/InputCSVParserV3.java:41]: could be but this works fine and reads much better.

File: app/services/InputCSVParserV3.java (hunk 3 above)
Comment (fenallen) [reply app/services/InputCSVParserV3.java:70]: Possibly - this is poc code and this can be looked at in the story to put it in the codebase 'for real'

**Review by YetAnotherMatt — APPROVED** [review]

**Pull request merged**